draft-ietf-tls-tls13.md

```diff
@@ -4696,6 +4642,195 @@ a context in all exporter computations, though the value could be empty.
+concerned with:
+
+- Network attackers who mount a replay attack by simply duplicating a
+  flight of 0-RTT data.  
+  
+- Network attackers who take advantage of client retry behavior
+  to arrange for the server to receive multiple copies of an application
+  message. This threat already exists
+  to some extent because browsers respond to network errors by
+  attempting to retry requests. However, 0-RTT adds an additional
+  dimension for any server system which does not maintain globally
+  consistent server state. Specifically, if a server system has
+  multiple zones where tickets from zone A will not be accepted in
+  zone B, then an attacker can duplicate a ClientHello and early
+  data intended for A to both A and B. At A, the data will
+  be accepted in 0-RTT, but at B the server will reject 0-RTT
+  data and instead force a full handshake. If the attacker, blocks
+  the ServerHello from A, then the client will complete the
+  handshake and retry the request, leading to duplication on
+  the server system as a whole.
+
+The first class of attack can be prevented by the mechanism described
+in this section.  Servers need not permit 0-RTT at all, but those
+which do SHOULD implement either the single-use tickets or
+ClientHello recording techniques described in the following two
+sections They MAY additionally implement time-based filtering
+{{time-based-filtering}} as an initial
+screen to prevent gross replays.
+
+The second class of attack cannot be prevented at the TLS layer and
+must be dealt with by any application. Note that any application whose
+clients implement any kind of retry behavior already needed to
+implement some sort of anti-repley defense.
+
+In normal operation, clients will not know which, if any, of these
+mechanisms servers actually implement and therefore MUST only send
+early data which they are willing to have subject to the attacks
+described in {{replay-0rtt}}.
+
+
+## Single-Use Tickets
+
+The simplest form of anti-replay defense is for the server to only
+allow each session ticket once. In order to implement this, the server
+maintains a database of all outstanding valid tickets; deleting each
+ticket from the database as it is used. If an unknown ticket is
+provided, the server falls back to a full handshake as normal.
+
+If the tickets are not self-contained but rather are database keys,
+and these PSKs are deleted upon use, then connections established
+using one PSK enjoy forward security. This is a security advantage for
+all 0-RTT data and for PSK usage when PSK is used without DH.
+
+Because this mechanism requires sharing the session database between
+server nodes in environments with multiple distributed servers,
+in such cases it may be hard to achieve high rates of PSK 0-RTT
+success when compared with self-encrypted tickets which do not
+require consistent server-side storage for PSK-based session
+establishment but do require it for anti-replay if 0-RTT is allowed,
+as described below.
```

Comment:
This is a very long sentence, and I started to lose the thread by the end.

> In an environment with multiple distributed servers, this mechanism requires sharing the session database between server nodes.  In these deployments it could be hard to achieve high rates of 0-RTT success when compared to self-encrypted tickets, which don't depend on consistent server-side storage.

I just don't know what you mean by the final clause: "but do require it for anti-replay".